Add Remoting Subproject

oleg-nenashev opened the pull request:
- [x] Add Remoting subproject page
- [x] Add `remoting` tag to the Remoting CLI deprecation post
- [x] Redirect links for https://github.com/jenkinsci/jenkins/pull/2950 ([JENKINS-45841](https://issues.jenkins-ci.org/browse/JENKINS-45841))

Screenshots:

![screen shot 2017-08-04 at 13 15 36](https://user-images.githubusercontent.com/3000480/28966535-06f13fea-7917-11e7-8077-606f06482e87.png)

![screen shot 2017-08-04 at 13 15 16](https://user-images.githubusercontent.com/3000480/28966540-0aefd28c-7917-11e7-9785-4d3cde2667d5.png)



@rtyler @reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented:
> Redirect links only make sense when linked to from somewhere we cannot arbitrarily update, like Jenkins itself. There's no reason to use those for links from the site. If the destination URL changes, we can just change it directly, rather than change the direct.

Maybe. I have just though that a "Redirecting to..." popup is useful, especially since I plan to move the docs to jenkins.io at some point. Can remove the links, but I see no problem with that.

> AFAICT jenkinsci/jenkins#2950 needs just two URLs, and those could easily be /redirect/remoting-jnlp3-deprecation and /redirect/remoting-jnlp2-deprecation, maybe adding one directory for remoting, but that's stretching it already.

I will likely use other links soon. In the core built-in documentation and Support Core. I would prefer to keep them for now. 

Regarding the redirect structure, as a remoting maintainer I prefer to use hierarchy within the subproject. It will simplify further management of the links. Do you block the links due to the structure I have selected? I see you do the same for Wiki links

oleg-nenashev commented:
@daniel-beck Just to follow-up on the discussion:

* Whatever redirects and pages are now stored in the Remoting subproject
* I have moved all redirects directly into the Remoting Project Directory
* Remoting project index page now uses relative links, hence there is no issue with explicit redirects. Once I setup Remoting docs publishing to Jenkins IO, the redirects will be just replaced by their pages.

rtyler commented on content/projects/index.adoc:
```diff
@@ -89,3 +89,16 @@ and operating the meetup group.
+
+image:/images/network-workgroup.png["Infrastructure', role=right]
+
+Jenkins Remoting is a library and executable JAR, which implements communication layer in Jenkins automation server: 
```
This paragraph is missing some articles, etc.

> Jenkins Remoting is a library, and executable Java archive, which implements the communication layer in Jenkins. This includes the TCP-based communication protocols, remote procedure calls, class loading, data streaming, etc. Currently Remoting is primarily used in master/agent communications, but is also used for the Remoting-based CLI and  the plugin:maven-plugin[Maven Integration plugin].
>
> The Remoting sub-project includes the Remoting library itself, package for agents, and a number of Remoting-specific plugins and core modules.

rtyler commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+tags:
+- remoting
+- agent
+- agents
```
Tags are not useful/used on pages like this. Only on blog posts.

oleg-nenashev replied:
Tags here are used to show the "Recent blog posts" tabs on the right side

rtyler commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+- slaves
+---
+
+Jenkins Remoting is a library and executable JAR, which implements communication layer in Jenkins: 
```
Replace with the suggested paragraph from the Sub-projects home-page

rtyler commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+== Remoting Usage in Jenkins
+
+* Agent executables (`slave.jar` is a renamed `remoting.jar`)
+* Master to Agent communication protocols being used in various Agent types, including Java Web Start (aka JNLP) and link:https://plugins.jenkins.io/ssh-slaves[SSH] agents 
```
link to plugins with `plugin:artifactId[Display Name here]`

rtyler commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+* Agent executables (`slave.jar` is a renamed `remoting.jar`)
+* Master to Agent communication protocols being used in various Agent types, including Java Web Start (aka JNLP) and link:https://plugins.jenkins.io/ssh-slaves[SSH] agents 
+* link:/doc/book/managing/cli/[Jenkins CLI], Remoting mode is **deprecated since 2.54** (link:/blog/2017/04/11/new-cli/[announcement])
+* Communication with Maven instance in the link:https://plugins.jenkins.io/maven-plugin[Maven Integration plugin]
```
`plugin:maven-plugin[Maven Integration plugin]`

rtyler commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+
+=== Remoting Library
+
+Remoting Library is a main component of the Remoting subproject.
```
s/is a main/is the main/

s/subproject/sub-project/

rtyler commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+=== Remoting Library
+
+Remoting Library is a main component of the Remoting subproject.
+The code base is located in the link:https://github.com/jenkinsci/remoting[separate repository],
```
s/in the/in a/

bitwiseman commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+- agent
+- agents
+- slave
+- slaves
```
The terms "slave" is deprecated.  Let's not continue to support it.

oleg-nenashev replied:
It is to aggregate links to old blogposts. But I agree it rather makes sense to update labels there

bitwiseman commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+
+== Remoting Usage in Jenkins
+
+* Agent executables (`slave.jar` is a renamed `remoting.jar`)
```
Perhaps: 
```
* Agent executables (`remoting.jar` is a named `slave.jar' in Jenkins for backwards compatibility)

bitwiseman commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+=== Jenkins Core and Modules
+
+Although Remoting library is a standalone component, 
+Jenkins core includes some extension and modules for working with Remoting.
```
extensions

bitwiseman commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+
+Although Remoting library is a standalone component, 
+Jenkins core includes some extension and modules for working with Remoting.
+It allows implementing custom protocols on-demand (e.g. for Jenkins CLI).
```
s/It/The Remoting Library/ 
or 
s/It/Jenkins/

bitwiseman commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+Although Remoting library is a standalone component, 
+Jenkins core includes some extension and modules for working with Remoting.
+It allows implementing custom protocols on-demand (e.g. for Jenkins CLI).
+There is ongoing activity on decoupling these components to plugins 
```
s/components to plugins/components from other Jenkins plugins/

oleg-nenashev replied:
"There is ongoing activity on detaching these core components to Jenkins plugins"

bitwiseman commented on content/projects/remoting/index.adoc:
```diff
@@ -0,0 +1,73 @@
+
+=== Packaging
+
+There are several packages, which bundle the Remoting library and allow connecting Jenkins agents
```
packages which

oleg-nenashev commented:
> The redirects are only acceptable if those pages are planned to be moved into jenkins.io.

This is my plan once we figure out how to publish docs from external sources. Originally I put them to "/redirect", but then I moved them to the project after the discussion with @daniel-beck

rtyler commented:
@oleg-nenashev "publish docs from external sources"

We already incorporate more external data than I am comfortable, either jenkins.io should be the canonical source for this documentation, or just link to GitHub. Please do not add additional external dependencies to this project just for pretty URLs :smiley_cat:

oleg-nenashev commented:
> We already incorporate more external data than I am comfortable, either jenkins.io should be the canonical source for this documentation, or just link to GitHub. Please do not add additional external dependencies to this project just for pretty URLs

Just a link to the GitHub documentation then. I insist on having Documentation-as-Code for the Remoting library. If the jenkins.io maintainers do not want to publish these docs on the website for simplified UX and Evangelism purposes, fine for me.

**Disclaimer:** The comment above represents my personal opinion. It may differ from official position of my company

rtyler commented:
> Just a link to the GitHub documentation then. I insist on having Documentation-as-Code for the Remoting library. If the jenkins.io maintainers do not want to publish these docs on the website for simplified UX and Evangelism purposes, fine for me.


My views have nothing to do with any "official position" and I'm annoyed at the implication. There are over a thousand repositories in the `jenkinsci` GitHub organization; external documentation for that many repositories is simply impractical.


The common argument for external documentation which I have heard is "putting the documentation near the code" because developers are *far* too busy to consult with multiple repositories (`</.sarcasm>`). Frankly, I think that's a load of crap. If that were the case remoting would be one big monolithic repo, as would many other multi-repository systems in the Jenkins project. 

There are absolutely no problems linking to external content from any pages within jenkins.io; many pages already do exactly that.  I don't think it makes any sense to add redirect pages to the URL hierarchy (outside of `redirects/`), or custom hacks just for downloading project-specific content into the site's build  just for vanity's sake.

daniel-beck commented:
@oleg-nenashev 

> Originally I put them to "/redirect", but then I moved them to the project after the discussion with @daniel-beck

FTR I never requested this. In fact I asked you repeatedly to move the 'remoting' folder back to 'redirects'.

oleg-nenashev commented:
> My views have nothing to do with any "official position" and I'm annoyed at the implication. 

Sorry if my comment annoys you. It has nothing to do with your opinion and with your comments. It is only a disclaimer for **my** comment, because I am not sure it reflects the position of my company && this PR has been created during my working hours. Hence there is such comment.

> I have heard is "putting the documentation near the code" because developers are far too busy to consult with multiple repositories 

They are, really. It complicates reviews of changes, and always leads to the "We will write the documentation later" cases. 

* Documentation versioning also becomes a PITA when you keep it separately. It is a real case for Remoting, where we may need to maintain LTS branches in order to reflect Jenkins LTS lines.
* I was maintaining Remoting 2.x for more than 6 month. In such case I would also need documentation versioning... if Remoting 2.x had documentation (excepting scattered Wiki pages)

>  I don't think it makes any sense to add redirect pages to the URL hierarchy (outside of redirects/), or custom hacks just for downloading project-specific content into the site's build just for vanity's sake.

I can just revert https://github.com/jenkins-infra/jenkins.io/pull/1056/commits/ff55900ee33e9ef5ff3f3ce1ced7b1906f0c09cf . Would it work for you?

oleg-nenashev commented:
@daniel-beck 

> FTR I never requested this. In fact I asked you repeatedly to move the 'remoting' folder back to 'redirects'.

If so, I probably misunderstood you. IIRC you were requesting having a flat structure of redirects, and then we switched to reverting Redirect links in the subproject change. I do not recall proposals about moving the redirects back (though I moved them to to /projects/remoting during the call for sure).

rtyler commented:
> I can just revert ff55900 . Would it work for you?

Yeah, that's fine by me. But why bother with redirects at all in this case? If the content is never going to move to jenkins.io, why not just make those links go directly to the resources in the git repository?

oleg-nenashev commented:
> Yeah, that's fine by me. But why bother with redirects at all in this case? If the content is never going to move to jenkins.io, why not just make those links go directly to the resources in the git repository?

These redirects will be used in https://github.com/jenkinsci/jenkins/pull/2950. Actually, all links excepting "troubleshooting". So I have created these links as a follow-up to this discussion:  https://github.com/jenkinsci/jenkins/pull/2950#discussion_r130375509

rtyler commented:
Sounds good to me

oleg-nenashev commented:
OK, will fix the bunch of comments tomorrow. Thanks for the reviews @daniel-beck, @rtyler and @bitwiseman
